doc/source/ray-core/examples/rdt/grpo_contextual_bandits.py

```diff
@@ -0,0 +1,470 @@
+        #   ratio < 1 - ε: the min() selects the unclipped branch. The model is penalized for reducing the probabily of a better action.
+
+        # For advantages < 0 (bad action):
+        #   ratio > 1 + ε: the min() selects the unclipped branch. The model is penalized for increasing the probabily of a bad action.
+        #   ratio < 1 - ε: the min() selects the clipped branch. This prevents the
+        #     policy from decreasing the probability for a bad action "too much".
+        loss = -torch.min(unclipped, clipped).mean()
+        # Fraction of actions which did not contribute to the gradient update.
+        clip_fraction = (
+            ((ratio < 1 - GRPO_CLIP_EPS) | (ratio > 1 + GRPO_CLIP_EPS)).float().mean()
+        )
+
+        # Update the policy network.
+        self.optim.zero_grad()
+        loss.backward()
+        # Clip the gradients to prevent exploding gradients and stabilize training.
+        nn.utils.clip_grad_norm_(self.model.parameters(), GRAD_CLIP_NORM)
+        self.optim.step()
+        self.scheduler.step()
+        self.policy_version += 1
+
+        return {
+            "loss": loss.detach().item(),
+            "clip_fraction": clip_fraction.detach().item()
+        }
+
+    def step(self) -> dict[str, Any]:
+        """Perform one training step and return metrics.
+
+        Each step samples a batch of trajectory slices from the replay buffer, computes the advantages, and updates the policy using the GRPO algorithm.
+        """
+        slices: list[TrajectorySlice] = ray.get(
+            self.replay_buffer.sample_from.remote(BATCH_SIZE)
+        )
+        # Prepare the tensors for the policy update.
+        actions = torch.cat([s["actions"] for s in slices]).to("cuda")
+        old_logps = torch.cat([s["old_logps"] for s in slices]).to("cuda")
+        rewards = torch.cat([s["rewards"] for s in slices]).to("cuda")
+        mean_rewards = torch.mean(rewards).item()
+        states = torch.stack([s["state"] for s in slices])
+        states = states.repeat_interleave(GROUP_SIZE, 0).to("cuda")
+
+        # Compute advantages and update the policy network using GRPO.
+        advantages = self._compute_advantages(rewards)
+        results = self._apply_policy_update(states, actions, old_logps, advantages)
+        results["rewards"] = mean_rewards
+
+        return results
+
+    @ray.method(tensor_transport="nixl")
+    def get_weights(self) -> dict[str, torch.Tensor]:
+        """Get the current model weights.
+
+        The tensor_transport="nixl" option enables NIXL via RDT to transfer model weight
+        tensors. Without it, the weights will be transferred using the Ray object store.
+        """
+        # Note: deepcopy() is needed because RDT does not support pointers to tensors yet.
+        # This will not be needed in the future.
+        state_dict = copy.deepcopy(self.model.state_dict())
```

Comment:
@Qiaolin-Yu should we link to a Github issue here?

Reply:
We can remove this now as Qiaolin found that the issue was actually storing the same tensor twice simultaneously.